### PR TITLE
Add user defaults monitor

### PR DESCRIPTION
**Pull request by Descartess**
#### What is the purpose of this pull request?
Add User Defaults Monitor

#### Background
#44 

#### Checklist

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to change).
- [x] I have read the **CONTRIBUTING** guidelines.
- [x] My code follows the coding style of this project.
- [x] My change requires a change to the documentation.
- [x] I have documented all my changes.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests pass.

### CODE REVIEW

**Comment by Descartess** [comment]
@eBardX I am blocked as the notifications parameter  throw an error when passed into the `observe` method. And i also don't have a clear idea on how to handle the notification object for the special cases

**Comment by eBardX** [comment]
@Descartess The `init` for `UserDefaultsMonitor` should have a `userDefault: UserDefaults` parameter. (The `.standard` instance of `UserDefaults` is just _one_ of many possible instances of the class that can be potentially monitored.) There should also be a `public let userDefault: UserDefaults` property on the monitor. For testing purposes, `UserDefaults()` _should_ be a legitimate instance.

**Comment by eBardX** [comment]
Sorry @Descartess! I hit the approve button prematurely.

**Comment by Descartess** [comment]
Yes I was wondering what had happened , but I am blocked the  Userdefaults notifications throw an error when  I pass the notification into the observe method

**Comment by eBardX** [comment]
I will try to pull your branch down and take a look tomorrow.

File: Sources/Core/Foundation/UserDefaultsMonitor.swift [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    private let handler: (Event) -> Void
+    private let options: Options
+
+    public override func addNotificationObservers() {
+        super.addNotificationObservers()
+        // Currently blocked as Xcode doesnt recognise `completedInitialCloudSyncNotification`
+        // I also check my closure implementation.
+        if options.contains(.completedInitialCloudSync) {
```
Comment (eBardX) [inline Sources/Core/Foundation/UserDefaultsMonitor.swift]: @Descartess This seems to work:
```
        #if os(iOS) || os(tvOS) || os(watchOS)
        if options.contains(.completedInitialCloudSync) {
            if #available(iOS 9.3, *) {
                observe(UserDefaults.completedInitialCloudSyncNotification) { [ unowned self] _ in
                    self.handler(.completedInitialCloudSync)
                }
            }
        }
        #endif
```

**Comment by Descartess** [comment]
@eBardX  Please review my current PR. Also pay special attention to the `sizeLimitExceeded` test I was not sure what exactly to assert, The current test only checks that the expectation was fulfilled.

File: Sources/Core/Foundation/UserDefaultsMonitor.swift [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+//  XestiMonitors
+//
+//  Created by Paul Nyondo on 2018-04-25.
+//
+//  © 2018 J. G. Pusey (see LICENSE.md)
+//
+
+import Foundation
+
+public class UserDefaultsMonitor: BaseNotificationMonitor {
```
Comment (eBardX) [inline Sources/Core/Foundation/UserDefaultsMonitor.swift]: There should be a doc comment for the class.

File: Sources/Core/Foundation/UserDefaultsMonitor.swift [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+    ///
+    /// Initializes a new `UserDefaultsMonitor`.
+    ///
+    /// - Parameters:
+    ///   - options:    The options that specify which events to monitor. By
+    ///                 default, all events are monitored.
+    ///   - queue:      The operation queue on which the handler executes. By
+    ///                 default, the main operation queue is used.
+    ///   - handler:    The handler to call when the current focus is updated
```
Comment (eBardX) [inline Sources/Core/Foundation/UserDefaultsMonitor.swift]: The `handler` comment looks like a cut-and-paste error.

File: Sources/Core/Foundation/UserDefaultsMonitor.swift [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    ///                 or cannot be moved in the selected direction.
+    ///
+    public init(options: Options = .all,
```
Comment (eBardX) [inline Sources/Core/Foundation/UserDefaultsMonitor.swift]: The user defaults instance should be passed in as a parameter.

File: Sources/Core/Foundation/UserDefaultsMonitor.swift [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        super.addNotificationObservers()
+
+        if options.contains(.didChange) {
+            observe(UserDefaults.didChangeNotification) { [unowned self] in
+                if let userDefaults = $0.object as? UserDefaults {
+                    self.handler(.didChange(userDefaults))
+                }
+            }
+        }
+        #if os(tvOS)
```
Comment (eBardX) [inline Sources/Core/Foundation/UserDefaultsMonitor.swift]: I believe this option applies to all platforms _except_ macOS.

File: Sources/Core/Foundation/UserDefaultsMonitor.swift [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    private let handler: (Event) -> Void
+    private let options: Options
+
+    public override func addNotificationObservers() {
+        super.addNotificationObservers()
+
+        if options.contains(.didChange) {
+            observe(UserDefaults.didChangeNotification) { [unowned self] in
```
Comment (eBardX) [inline Sources/Core/Foundation/UserDefaultsMonitor.swift]: Should only observe this notification on the passed in user defaults object.

File: Tests/Foundation/UserDefaultsMonitorTests.swift [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+//
+//  © 2018 J. G. Pusey (see LICENSE.md)
+//
+
+import XCTest
+@testable import XestiMonitors
+
+internal class UserDefaultsMonitorTests: XCTestCase {
+    let notificationCenter = MockNotificationCenter()
+    let userDefaults = UserDefaults.standard
```
Comment (eBardX) [inline Tests/Foundation/UserDefaultsMonitorTests.swift]: I think the tests will be better if you use `UserDefaults()` to create a brand new (non-shared) instance.

File: Tests/Foundation/UserDefaultsMonitorTests.swift [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+                                          queue: .main) { event in
+                                            expectedEvent = event
+                                            expectation.fulfill()
+        }
+
+        monitor.startMonitoring()
+        simulateSizeLimitsExceeded()
+        waitForExpectations(timeout: 1)
+        monitor.stopMonitoring()
+    }
```
Comment (eBardX) [inline Tests/Foundation/UserDefaultsMonitorTests.swift]: You should also test that `expectedEvent` equals `.sizeLimitExceeded`.

File: Tests/Foundation/UserDefaultsMonitorTests.swift [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+        if let event = expectedEvent,
+            case let .didChange(test) = event {
+            XCTAssertEqual(test, userDefaults)
+        } else {
+            XCTFail("Unexpected event")
+        }
+    }
+
+    func testMonitor_sizeLimitsExceeded() {
```
Comment (eBardX) [inline Tests/Foundation/UserDefaultsMonitorTests.swift]: Should be `testMonitor_sizeLimitExceeded` (singular `Limit`).

File: Tests/Foundation/UserDefaultsMonitorTests.swift [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+        simulateSizeLimitsExceeded()
+        waitForExpectations(timeout: 1)
+        monitor.stopMonitoring()
+    }
+
+    private func simulateDidChange() {
+        notificationCenter.post(name: UserDefaults.didChangeNotification, object: userDefaults)
+    }
+
+    private func simulateSizeLimitsExceeded() {
```
Comment (eBardX) [inline Tests/Foundation/UserDefaultsMonitorTests.swift]: Should be `simulateSizeLimitExceeded` (singular `Limit`).

File: Tests/Foundation/UserDefaultsMonitorTests.swift [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        notificationCenter.post(name: UserDefaults.sizeLimitExceededNotification, object: nil)
```
Comment (eBardX) [inline Tests/Foundation/UserDefaultsMonitorTests.swift]: Have we determined definitively that `UserDefaults.sizeLimitExceededNotification` does _not_ have a notification object? I suspect that it is highly likely that it also users a `UserDefaults` instance for the notification object even though the docs are silent on the matter.

File: Sources/Core/Foundation/UserDefaultsMonitor.swift (hunk 6 above)
Comment (Descartess) [reply Sources/Core/Foundation/UserDefaultsMonitor.swift]: `Static member 'didChangeNotification' cannot be used on instance of type 'UserDefaults`' I get this error when I try to get to access the notification from the UserDefaults object

**Comment by eBardX** [comment]
@Descartess For some obscure reason the tests are failing on the tvOS 9.0 simulator only. Please do _not_ merge this PR until we get to the bottom of it. I'll continue looking into it. Feel free to do likewise.

**Pull request merged**